core/include/edm/collection.hpp

```diff
@@ -0,0 +1,180 @@
+
+    /// @}
+
+    /**
+     * @brief The size type of this container, which is the type by which
+     * its elements are indexed.
+     */
+    using size_type = typename item_vector::size_type;
+
+    /**
+     * @brief Standard constructor.
+     *
+     */
+    template <typename item_vector_tp,
+              typename = std::enable_if<
+                  std::is_same<item_vector_tp, vector_t<item_t>>::value>>
+    TRACCC_HOST_DEVICE collection(item_vector_tp&& iv) : items(iv) {}
+
+    /**
+     * @brief Constructor with vector and memory resource .
+     */
+    TRACCC_HOST collection(item_vector iv, vecmem::memory_resource* mr)
+        : items(iv, mr) {}
+
+    /**
+     * @brief Constructor with memory resource .
+     */
+    TRACCC_HOST explicit collection(vecmem::memory_resource* mr) : items(mr) {}
+
+    /**
+     * @brief Default Constructor
+     */
+    collection() = default;
+
+    /**
+     * @brief Mutable element accessor.
+     */
+    TRACCC_HOST_DEVICE
+    item_t& operator[](size_type i) { return items.at(i); }
+
+    /**
+     * @brief Immutable element accessor.
+     */
+    TRACCC_HOST_DEVICE
+    const item_t& operator[](size_type i) const { return items.at(i); }
+
+    /**
+     * @brief Return the size of the collection.
+     */
+    TRACCC_HOST_DEVICE
+    size_type size(void) const { return items.size(); }
+
+    /**
+     * @brief Reserve space in both vectors.
+     */
+    TRACCC_HOST
+    void reserve(size_type s) { items.reserve(s); }
+
+    /**
+     * @brief Non-const accessor method for the internal header vector.
+     *
+     * @warning Do not use this function! It is dangerous, and risks breaking
+     * invariants!
+     */
+    TRACCC_HOST_DEVICE
+    item_vector& get_items() { return items; }
+
+    /**
+     * @brief Push an element into the collection.
+     */
+    template <typename i_prime>
+    TRACCC_HOST void push_back(i_prime&& new_item) {
+        items.push_back(std::forward<item_t>(new_item));
+    }
+
+    TRACCC_HOST_DEVICE
+    auto begin() { return items.begin(); }
+
+    TRACCC_HOST_DEVICE
+    auto end() { return items.end(); }
+
+    /// All objects in the event
+    item_vector items;
+};
+
+/// Convenience declaration for the collection type to use in host code
+template <typename item_t>
+using host_collection = collection<item_t, vecmem::vector>;
+
+/// Convenience declaration for the collection type to use in device code
+template <typename item_t>
+using device_collection = collection<item_t, vecmem::device_vector>;
+
+/// @name Types used to send data back and forth between host and device code
+/// @{
+
+/// Structure holding (some of the) data about the collection in host code
+template <typename item_t>
+struct collection_data {
+    vecmem::data::vector_view<item_t> items;
+};
+
+/// Structure holding (all of the) data about the collection in host code
+template <typename item_t>
+struct collection_buffer {
+    vecmem::data::vector_buffer<item_t> items;
+};
```

Comment:
I should have looked carefully at #75, sorry for not doing that. :frowning:

I don't see the point in this class... :confused: What does `collection` give us over:

```c++
template <typename item_t>
using host_collection = vecmem::vector<item_t>;

template <typename item_t>
using device_collection = vecmem::device_vector<item_t>;
```

? It literally just wraps an existing vector type, without adding any new feature to it (as far as I can tell...).

The same is true for the `collection_data` and `collection_buffer` types. What sort of abstraction do these give to the traccc code over using the vecmem types directly?

Reply:
No advantage really. I just wanted to keep the same behavior of container class for vector-based class as well. Of course it is not mandatory thing so If you and @stephenswat  don't like it, I can remove this class and close PR75

Reply:
In my opinion, since the work is already there, I don't see any reason to remove it. If it works, it works. I might not completely understand why the code is there exactly but you're the one writing this algorithm so if your judgement is that this is helpful then I am personally happy to go along with it.

Reply:
In this case I fear that this design unnecessarily complicates things. As this intermediate layer between the vecmem types and the traccc algorithmic code does not add any features as far as I can see.

The `traccc::container` type is quite different in this respect. That does add a non-trivial amount of logic on top of the vecmem types, so that one I really like... I just feel that this type just makes the code a little harder to read... 🤔

Reply:
I agree. It apparently looks better after removing collection class. I will close PR75